### PR TITLE
[V1][Spec Decode] Optimize Rejection Sampler with Triton Kernels

### PR DESCRIPTION
**Pull request by WoosukKwon**
This PR optimizes the rejection sampler in #13933 with custom Triton kernels. 

By using the Triton kernels, the PR brings the following benefits:
1. Now we use the flattened shape `[num_tokens, vocab_size]` for the logits tensors, instead of `[batch_size, max_spec_len, vocab_size]`. This reduces the GPU memory usage a lot.
2. Zero synchronization between CPU and GPU.
3. Remove inefficient data movement (i.e., a bunch of `cat`, `gather`, etc.)
4. (Arguably) easier-to-read code

### Performance benchmark: Llama 3.1 8B, ShareGPT, 1xH100, temperature 0.1
SD config: `--speculative-model "[ngram]" --ngram_prompt_lookup_min 5 --ngram-prompt-lookup-max 5 --num_speculative_tokens 3`

Screenshot 2025-03-17 at 11 28 14 AM

| | Throughput (reqs/s) |
|----------|:----------:|
| main (w/o SD) | 51.49 |
| main (w/ SD) | 54.41 |
| This PR (w/ SD) | 64.16 |

25% throughput increase compared to main w/o SD, and 18% increase compared to main w/ SD.

### Accuracy benchmark: GSM8K, Llama 3.1 8B Instruct, 5 shots

| | Temperature | Exact match |
|----------|:----------:|:----------:|
| w/o SD | 0.0 | 75.7 |
| | 1.0 | 50.9 |
| w/ SD | 0.0 | 75.9 |
| | 1.0 | 51.8|

### CODE REVIEW

File: vllm/v1/sample/rejection_sampler.py [hunk 1]
```diff
@@ -31,317 +38,521 @@ class RejectionSampler(nn.Module):
...
+
+    vocab_offset = tl.arange(0, PADDED_VOCAB_SIZE)
+    if IS_NGRAM:
+        draft_token_id = tl.load(draft_token_ids_ptr + start_idx + pos)
+        orig_prob = tl.load(target_probs_ptr + (start_idx + pos) * vocab_size +
+                            draft_token_id)
+        # Temporarily zero out the probability of the draft token.
+        tl.store(
+            target_probs_ptr + (start_idx + pos) * vocab_size + draft_token_id,
+            0)
```
Comment (LiuXiaoxuanPKU) [inline vllm/v1/sample/rejection_sampler.py:530]: QQ: why do we need to zero out prob here?

**Review by LiuXiaoxuanPKU — COMMENTED** [review]
Finished the rejection_sampler.py, will continue other files tonight

File: vllm/v1/sample/rejection_sampler.py [hunk 2]
```diff
@@ -31,317 +38,521 @@ class RejectionSampler(nn.Module):
...
-    batch_size, num_tokens = token_ids.shape
+    """Compute probability distribution from logits based on sampling metadata.
 
-    token_probs = torch.zeros(batch_size,
-                              num_tokens,
-                              vocab_size,
-                              dtype=torch.float,
-                              device=out_device)
+    This function applies temperature scaling to the logits and converts 
+    them to probabilities using softmax. For greedy decoding
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/v1/sample/rejection_sampler.py:228]: Nice catch. Fixed.

File: vllm/v1/sample/rejection_sampler.py [hunk 3]
```diff
@@ -31,317 +38,521 @@ class RejectionSampler(nn.Module):
...
+    assert draft_token_ids.is_contiguous()
+    assert draft_probs is None or draft_probs.is_contiguous()
+    assert target_probs.is_contiguous()
+    assert bonus_token_ids.is_contiguous()
+    assert target_probs.shape == (num_tokens, vocab_size)
+
+    # Create output buffer.
+    output_token_ids = torch.empty(
+        (batch_size, max_spec_len + 1),
+        dtype=torch.int64,
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/v1/sample/rejection_sampler.py:154]: Nice catch! Fixed.

File: vllm/v1/sample/rejection_sampler.py (hunk 1 above)
Comment (WoosukKwon) [reply vllm/v1/sample/rejection_sampler.py:530]: Basically, the recovered token from N-gram is `argmax(target_prob)` except the draft token id.
This is essentially the same as `target_prob - draft_prob` below, except that n-gram does not have draft_prob. We regard it as 1.

Added a comment explaining this.

**Review by LiuXiaoxuanPKU — APPROVED** [review]
LGTM, thanks!

File: vllm/v1/sample/ops/utils.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    torch._dynamo.mark_dynamic(logits, index=0)
+    if isinstance(temperature, torch.Tensor):
+        torch._dynamo.mark_dynamic(temperature, index=0)
+    return _softmax(logits, temperature)
+
+
+@torch.compile
+def _softmax(
+    logits: torch.Tensor,
+    temperature: Union[float, torch.Tensor],
```
Comment (youkaichao) [inline vllm/v1/sample/ops/utils.py:27]: some late comments: if `temperature` is float, then it will still recompile when `temperature` changes.

Comment (WoosukKwon) [reply vllm/v1/sample/ops/utils.py:27]: @youkaichao thanks for the heads up. After this PR, I find that this softmax doesn't show noticeably better performance than pytorch's native softmax. I will remove this.

Comment (WoosukKwon) [reply vllm/v1/sample/ops/utils.py:27]: Removed in #15416

File: vllm/v1/sample/rejection_sampler.py [hunk 5]
```diff
@@ -3,25 +3,32 @@
...
+from vllm.v1.spec_decode.metadata import SpecDecodeMetadata
 
 logger = init_logger(__name__)
-INVALID_TOKEN_ID = -1
+
+PLACEHOLDER_TOKEN_ID: tl.constexpr = -1
+GREEDY_TEMPERATURE: tl.constexpr = -1
+# Maximum number of speculative draft tokens allowed per request in a single
+# step. This value is chosen to be large enough to handle typical use cases.
+MAX_SPEC_LEN = 32
```
Comment (mmyxym) [inline vllm/v1/sample/rejection_sampler.py:20]: Hi @WoosukKwon , is there any limitation MAX_SPEC_LEN should be 32? Can it be larger? Thanks.

Comment (WoosukKwon) [reply vllm/v1/sample/rejection_sampler.py:20]: @mmyxym There's no blocker to make it 64. Everything should work if you just change the number. I just thought 32 would be enough for all practical use cases.
